### ISSUE TITLE
[Feature] Swappable renderer

### ISSUE THREAD

**Issue by NullVoxPopuli** [emberjs/ember.js#20648]
Based on how much struggle I've been having with the VM latetly:
- https://github.com/glimmerjs/glimmer-vm/pull/1563
- https://github.com/glimmerjs/glimmer-vm/pull/1547
- https://github.com/emberjs/ember.js/issues/20644
  - https://github.com/emberjs/ember.js/issues/20643
  - https://github.com/emberjs/ember.js/issues/20642 
- https://github.com/emberjs/ember.js/issues/20631
- https://github.com/emberjs/ember.js/issues/20610
- https://github.com/emberjs/ember.js/issues/20535

_And_ with the amount of progressing happening on an alternate renderer: 
- compat todos are done: https://github.com/lifeart/glimmer-next/issues/10 (we'll see tho if it works with ember-modifier, ember-resources, and bigger projects)
- faster reactivity technique: https://github.com/lifeart/glimmer-next/issues/41 (Update-only VM)

_And_ with how we don't score whell in https://github.com/krausest/js-framework-benchmark (like it or not, this is how we are judged, so we must play the game)

_And_ with the ovelwhelming effort it feels like it would take to merge the 3 repos:
- this one: https://github.com/emberjs/ember.js
- glimmer-vm: https://github.com/glimmerjs/glimmer-vm
- glimmer.js: https://github.com/glimmerjs/glimmer.js 
(which honestly needs to happen anyway)

I've been wondering if we can work out a minimal public/intimate API for completely swapping out glimmer-vm / runtime such that folks can better experiment with improvements to ember.

As a quickly, barely thought out psuedo API, it could look like this, to support what we have today
```js
import { registerRenderer } from '@ember/renderer';
// idk if these are real, but not the point
import { render } from '@glimmer/vm';
import { runtime } from '@glimmer/runtime';

registerRenderer({ render, runtnime });
```

which could allow for someone to easily choose to use `glimmer-next` without forking ember-source.

Or could allow someone to work on developing a _native_ renderer for iOS and Android.

------------

One may consider that we have a lot to do, and we don't have enough steam to work on anything like this (like a native renderer).

I think we have people who currently aren't contributing as much, but _would_ if they were able to, so we could enhance our parallel efforts -- widen the pipe, so to speak

**Comment by NullVoxPopuli** [comment]
Starbeam's vanilla renderer would be a good candidate here as well: https://github.com/starbeamjs/starbeam/issues/152

**Comment by NullVoxPopuli** [comment]
Another use case: Terminal UI renderers -- similar to Ink https://github.com/vadimdemedes/ink 
(I bet we could use way less memory pressure tho)

**Comment by NullVoxPopuli** [comment]
Looks like the place to implemetn this is in `setupApplicationRegistry` here:

https://github.com/emberjs/ember.js/blob/main/packages/%40ember/-internals/glimmer/lib/setup-registry.ts#L41
```js
  registry.register('renderer:-dom', Renderer);
```

This is called from `buildRegistry` here: 
https://github.com/emberjs/ember.js/blob/main/packages/%40ember/application/index.ts#L226

And that's called from `init` in the Application's suprer class, Engine, here: 
https://github.com/emberjs/ember.js/blob/main/packages/%40ember/engine/index.ts#L335
```ts
  init(properties: object | undefined) {
    super.init(properties);
    this.buildRegistry();
  }
```

And we know application init occurs during boot with the Application class we create in app/app.js/ts.

So, we should be able to override the `static buildRegistry` and setup on our own `renderer:-dom`.

**Comment by NullVoxPopuli** [comment]
woah, look at all the renderers React has: https://github.com/chentsulin/awesome-react-renderer

there is certainly demand for this.

**Comment by NullVoxPopuli** [comment]
Lynx: https://github.com/lynx-family/lynx -- native app building tools

**Comment by johanrd** [comment]
Also very inspiring to read https://mainmatter.com/blog/2025/05/22/native-apps-with-svelte/

**Comment by NullVoxPopuli** [comment]
Terminal framework with ember here by @patricklx  https://github.com/patricklx/ember-tui

**Comment by johanrd** [comment]
yes!, also intriguing: https://x.com/bcardarella/status/1999195518770041196, 
AND of course your talk at emberfest! https://www.youtube.com/watch?v=1JxHMw23Jmg
